File: Programmers/[81303] 표 편집.java
package pg81303;

import java.util.Stack;

class Solution {
    static Node initialList(int n, int k) {
        Node curNode = new Node(-1); // head

        Node newNode;
        for (int i = 0; i < n; i++) {
            newNode = new Node(i);
            curNode.down = newNode;
            newNode.up = curNode;
            curNode = newNode;
        }

        newNode = new Node(-1); // tail
        curNode.down = newNode;
        newNode.up = curNode;

        while (curNode.index != k) {
            curNode = curNode.up;
        }
        return curNode;
    }

    public String solution(int n, int k, String[] cmd) {
        Node curNode = initialList(n, k);
        Stack<Node> stack = new Stack<>();
        int number;

        for (String s : cmd) {
            switch (s.charAt(0)) {
                case 'U':
                    number = Integer.parseInt(s.substring(2));
                    for (int i = 0; i < number; i++) {
                        curNode = curNode.up;
                    }
                    break;
                case 'D':
                    number = Integer.parseInt(s.substring(2));
                    for (int i = 0; i < number; i++) {
                        curNode = curNode.down;
                    }
                    break;
                case 'C':
                    stack.add(curNode);
                    curNode.remove();
                    if (curNode.canMoveDown()) {
                        curNode = curNode.down;
                    } else {
                        curNode = curNode.up;
                    }
                    break;
                case 'Z':
                    stack.pop().restore();
                    break;
            }
        }

        StringBuilder answer = new StringBuilder("O".repeat(n));
        while (!stack.isEmpty()) {
            answer.setCharAt(stack.pop().index, 'X');
        }
        return answer.toString();
    }

    static class Node {
        int index;
        Node up, down;

        public Node(int index) {
            this.index = index;
        }

        boolean canMoveDown() {
            return down.index != -1;
        }

        void remove() {
            up.down = down;
            down.up = up;
        }

        void restore() {
            up.down = this;
            down.up = this;
        }
    }
}

File: Programmers/[81303] 표 편집.py
# 5 ≤ n ≤ 1,000,000
# 0 ≤ k < n
# 1 ≤ cmd의 원소 개수 ≤ 200,000


def solution(n, k, cmd):
    linked_list = [Node(i) for i in range(n)]
    for i in range(n - 1):
        linked_list[i].down = linked_list[i + 1]
        linked_list[i + 1].up = linked_list[i]
    linked_list[n - 1].up = linked_list[n - 2]

    # cursor
    node = linked_list[k]

    stack = []
    for command in cmd:
        if command[0] == "U":
            index = int(command.split()[1])
            for _ in range(index):
                node = node.up
        elif command[0] == "D":
            index = int(command.split()[1])
            for _ in range(index):
                node = node.down
        elif command[0] == "C":
            stack.append(node)
            node.is_deleted = True
            if node.up is not None:
                node.up.down = node.down
            if node.down is not None:
                node.down.up = node.up

            if node.down is not None:
                node = node.down
            else:
                node = node.up
        elif command[0] == "Z":
            popped = stack.pop()
            popped.is_deleted = False
            if popped.up is not None:
                popped.up.down = popped
            if popped.down is not None:
                popped.down.up = popped

    answer = ""
    for i in range(n):
        if linked_list[i].is_deleted:
            answer += "X"
        else:
            answer += "O"
    return answer


class Node:
    def __init__(self, number):
        self.number = number
        self.down = None
        self.up = None
        self.is_deleted = False
